File: backend/voiceGather/questionnaire/core/schema.py
import datetime
from dataclasses import asdict, dataclass, field
from typing import Any, List, Union

from questionnaire.core.extensions.enums import QuestionType


class BaseSchema:
    """
    ベーススキーマ
    """

    def as_dict(self):
        return asdict(self)


@dataclass
class CategorySummary(BaseSchema):
    """
    カテゴリスキーマ
    """

    id: int
    name: str
    count: int = None


@dataclass
class KeywordSummary(BaseSchema):
    """
    キーワードスキーマ
    """

    id: int
    keyword: str
    count: int = 0
    fast_rising: int = 0


@dataclass
class Questionnaire(BaseSchema):
    """
    アンケートスキーマ
    """

    id: int
    title: str
    overview: str
    answer_count: int
    scope: int
    start_at: datetime.date
    end_at: datetime.date
    created_at: datetime.datetime
    categories: List[CategorySummary]


@dataclass
class QuestionnaireMeta(BaseSchema):
    """
    アンケートメタデータスキーマ
    """

    id: int
    questionnaire_id: int
    index: int
    sub_index: int
    question_type: QuestionType


@dataclass
class Selection(BaseSchema):
    """
    選択肢スキーマ
    """

    number: int
    content: str


@dataclass
class BaseQuestion(BaseSchema):
    """
    設問のベーススキーマ
    """

    id: int
    questionnaire_id: int
    index: int
    is_subquestion: bool
    question_type: QuestionType


@dataclass
class QuestionFreeText(BaseQuestion):
    """
    フリーテキスト設問スキーマ
    """

    question: str
    answer_max_length: int
    required: bool
    answer_count: int


@dataclass
class QuestionSelect(BaseQuestion):
    """
    選択式設問スキーマ
    """

    question: str
    required: bool
    multi_select: bool
    selection: List[Selection]


@dataclass
class QuestionContainer(BaseSchema):
    """
    設問のコンテナ
    """

    id: int
    title: str
    start_at: datetime.date
    end_at: datetime.date
    data: List[BaseQuestion] = field(default_factory=list)


@dataclass
class BaseAnswerData(BaseSchema):
    """
    回答内容のベーススキーマ
    """

    answer_uuid: str
    parent_answer_uuid: str
    created_at: str


@dataclass
class FreeTextData(BaseAnswerData):
    """
    フリーテキストの回答内容スキーマ
    """

    answer: List[str]
    sub_answer: Union["FreeTextData", "SelectData"] = None


@dataclass
class SelectData(BaseAnswerData):
    """
    選択式回答内容スキーマ
    """

    answer: List[Selection]
    sub_answer: Union["FreeTextData", "SelectData"] = None


@dataclass
class TextSummary(BaseSchema):
    """
    フリーテキストの回答内容サマリスキーマ
    """

    count: int = 0


@dataclass
class SelectSummary(BaseSchema):
    """
    選択式の回答内容サマリスキーマ
    """

    number: int
    content: str
    count: int = 0


@dataclass
class BaseAnswer(BaseSchema):
    """
    回答のベーススキーマ
    """

    question_id: int
    question_type: QuestionType


@dataclass
class AnswerFreeText(BaseAnswer):
    """
    フリーテキスト回答スキーマ
    """

    data: List[FreeTextData]
    summary: TextSummary


@dataclass
class AnswerSelect(BaseAnswer):
    """
    選択式回答スキーマ
    """

    data: List[SelectData]
    summary: List[SelectSummary]

File: frontend/voice-gather/src/features/api/Schema.ts
export type Questionnaire = {
    id: number;
    title: string;
    overview: string;
    answerCount: number;
    startAt: string;
    endAt: string;
    categories: Array<Category>;
    createdAt: Date;
};

export type QuestionContainer = {
    id: number;
    title: string;
    start_at: string;
    end_at: string;
    data: Array<TextQuestion | SelectQuestion>;
}

export type TextQuestion = {
    id: number;
    questionnaire_id: string;
    index: number;
    is_subquestion: boolean;
    question_type: number;
    question: string;
    answer_max_length: number;
    required: boolean;
    answer_count: number;
}

export type SelectQuestion = {
    id: number;
    questionnaire_id: string;
    index: number;
    is_subquestion: boolean;
    question_type: number;
    question: string;
    required: boolean;
    multi_select: boolean;
    selection: Array<Selection>;
}

export type Selection = {
    number: number;
    content: string;
}

export type TextAnswer = {
    question_id: string;
    question_type: '1'; // ディスクリミネータ
    data: Array<TextAnswerData> | null;
    summary: TextSummary | null;
}

export type TextAnswerData = {
    answer_uuid: string;
    parent_answer_uuid: string;
    created_at: string;
    answer: Array<string>;
    sub_answer: TextAnswerData | SelectAnswerData | null;
}

export type TextSummary = {
    count: number;
}

export type SelectAnswer = {
    question_id: string;
    question_type: '2'; // ディスクリミネータ
    data: Array<SelectAnswerData> | null;
    summary: Array<SelectSummary> | null;
}

export type SelectAnswerData = {
    answer_uuid: string;
    parent_answer_uuid: string;
    created_at: string;
    answer: Array<number>;
    sub_answer: TextAnswerData | SelectAnswerData | null;
}

export type SelectSummary = {
    number: number;
    content: string;
    count: number;
}

export type Category = {
    id: number;
    name: string;
    count: number | null;
};

export type Keyword = {
    id: number;
    keyword: string;
    count: number | null;
    fastRising: number | null;
};